File: tests/lab3/test_cache.py
#tests/lab3/test_cache.py

import pytest
import os
from steps.test_steps import check_file_exists

CONSTRUCT_CACHE_SUCCESS = "[TEST 1] Construct cache: PASS"
CACHE_WRITE_SUCCESS = "[TEST 2] Cache write: PASS"
CACHE_READ_EXISTING_SUCCESS = "[TEST 3] Cache read (existing key): PASS"
CACHE_READ_MISSING_SUCCESS = "[TEST 4] Cache read (missing key): PASS"
CACHE_DESTRUCT_SUCCESS = "[TEST 5] Cache destruct: PASS"

@pytest.mark.dependency()
def test_cache_exists(project_dir):
    """Check that the test executable for cache exists."""
    cache_test = os.path.join(project_dir, "install", "test_cache")
    check_file_exists(cache_test)

@pytest.mark.dependency(depends=["tests/lab3/test_cache.py::test_cache_exists"])
def test_cache_basic(proxy_fixture, project_dir):
    """
    Test basic cache functionality:
    - construct_cache does not crash
    - cache_write successfully writes a key-value pair
    - cache_read successfully reads an existing key
    - Reading a missing key returns NULL
    - destruct_cache works correctly
    """
    cache_test = os.path.join(project_dir, "install", "test_cache")
    result = proxy_fixture.build_and_run_test(cache_test)
    stdout = result.stdout
    stderr = result.stderr

    assert CONSTRUCT_CACHE_SUCCESS in stdout, "construct_cache failed."
    assert CACHE_WRITE_SUCCESS in stdout, "cache_write did not succeed."
    assert CACHE_READ_EXISTING_SUCCESS in stdout, "cache_read failed for an existing key."
    assert CACHE_READ_MISSING_SUCCESS in stdout, "cache_read did not return NULL for a missing key."
    assert CACHE_DESTRUCT_SUCCESS in stdout, "destruct_cache failed."

File: tests/lab3/c_tests/test_cache.c
//tests/lab3/c_tests/test_cache.c

#include <stdio.h>
#include <string.h>
#include <assert.h>
#include "cache.h"

#define CACHE_SIZE (1024 * 1024)
#define KEY1 "key1"
#define VALUE1 "value1"
#define VALUE1_SIZE (strlen(VALUE1) + 1)

void test_construct_cache() {
    if (construct_cache(CACHE_SIZE) == 0) {
        printf("[TEST 1] Construct cache: PASS\n");
    } else {
        printf("[TEST 1] Construct cache: FAIL\n");
    }
}

void test_cache_write() {
    if (cache_write(KEY1, VALUE1, VALUE1_SIZE, 0) == 0) {
        printf("[TEST 2] Cache write: PASS\n");
    } else {
        printf("[TEST 2] Cache write: FAIL\n");
    }
}

void test_cache_read_existing_key() {
    size_t value_size;
    char *value = (char*)cache_read(KEY1, &value_size);
    if (value != NULL && strcmp(value, VALUE1) == 0 && value_size == VALUE1_SIZE) {
        printf("[TEST 3] Cache read (existing key): PASS\n");
    } else {
        printf("[TEST 3] Cache read (existing key): FAIL\n");
    }
}

void test_cache_read_missing_key() {
    size_t value_size;
    char *value = (char*)cache_read("unknown_key", &value_size);
    if (value == NULL) {
        printf("[TEST 4] Cache read (missing key): PASS\n");
    } else {
        printf("[TEST 4] Cache read (missing key): FAIL\n");
    }
}

void test_destruct_cache() {
    destruct_cache();
    printf("[TEST 5] Cache destruct: PASS\n");
}

int main() {
    printf("Running cache tests...\n");

    test_construct_cache();
    test_cache_write();
    test_cache_read_existing_key();
    test_cache_read_missing_key();
    test_destruct_cache();

    printf("All cache tests completed.\n");
    return 0;
}
